csharp/src/Ice/Communicator-ObjectAdapterFactory.cs

```diff
@@ -13,58 +14,48 @@ public sealed partial class Communicator
     {
         /// <summary>Shuts down this communicator's server functionality. This triggers the deactivation of all
         /// object adapters. After this method returns, no new requests are processed. However, requests that have
-        /// been started before Shutdown was called might still be active. You can use <see cref="WaitForShutdown"/>
-        /// to wait for the completion of all requests.</summary>
-        public void Shutdown()
+        /// been started before ShutdownAsync was called might still be active until the returned task completes. You
+        /// can also await the task returned by <see cref="WaitForShutdownAsync"/> to wait for the completion of all
+        /// requests.</summary>
+        public async Task ShutdownAsync()
         {
-            ObjectAdapter[] adapters;
             lock (_mutex)
             {
-                // Ignore shutdown requests if the communicator is already shut down.
-                if (_isShutdown)
-                {
-                    return;
-                }
-
-                adapters = _adapters.ToArray();
-                _isShutdown = true;
-                Monitor.PulseAll(_mutex);
+                _shutdownTask ??= PerformShutdownAsync();
             }
 
-            // Deactivate outside the lock to avoid deadlocks.
-            foreach (ObjectAdapter adapter in adapters)
+            try
+            {
+                await _shutdownTask.ConfigureAwait(false);
+            }
+            finally
             {
-                adapter.Deactivate();
+                _shutdownTaskSemaphore?.Release();
```

Comment:
mm should be a positive number greater than 0 otherwise it raises ArgumentException, so:

`_shutdownTaskSemaphore.Release(_shutdownTaskSemaphore.CurrentCount)`

https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim.release?view=netcore-3.1#System_Threading_SemaphoreSlim_Release_System_Int32_

Reply:
I replaced the Semaphore by a TaskCompletionSource seems more appropriate here